File: cpp/src/parquet/file/writer.cc
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

#include "parquet/file/writer.h"

#include "parquet/file/writer-internal.h"
#include "parquet/util/memory.h"

using parquet::schema::GroupNode;

namespace parquet {

// ----------------------------------------------------------------------
// RowGroupWriter public API

RowGroupWriter::RowGroupWriter(std::unique_ptr<Contents> contents)
    : contents_(std::move(contents)) {}

void RowGroupWriter::Close() {
  if (contents_) {
    contents_->Close();
  }
}

ColumnWriter* RowGroupWriter::NextColumn() { return contents_->NextColumn(); }

int RowGroupWriter::current_column() { return contents_->current_column(); }

int RowGroupWriter::num_columns() const { return contents_->num_columns(); }

int64_t RowGroupWriter::num_rows() const { return contents_->num_rows(); }

// ----------------------------------------------------------------------
// ParquetFileWriter public API

ParquetFileWriter::ParquetFileWriter() {}

ParquetFileWriter::~ParquetFileWriter() {
  try {
    Close();
  } catch (...) {
  }
}

std::unique_ptr<ParquetFileWriter> ParquetFileWriter::Open(
    const std::shared_ptr<::arrow::io::OutputStream>& sink,
    const std::shared_ptr<GroupNode>& schema,
    const std::shared_ptr<WriterProperties>& properties,
    const std::shared_ptr<const KeyValueMetadata>& key_value_metadata) {
  return Open(std::make_shared<ArrowOutputStream>(sink), schema, properties,
              key_value_metadata);
}

std::unique_ptr<ParquetFileWriter> ParquetFileWriter::Open(
    const std::shared_ptr<OutputStream>& sink,
    const std::shared_ptr<schema::GroupNode>& schema,
    const std::shared_ptr<WriterProperties>& properties,
    const std::shared_ptr<const KeyValueMetadata>& key_value_metadata) {
  auto contents = FileSerializer::Open(sink, schema, properties, key_value_metadata);
  std::unique_ptr<ParquetFileWriter> result(new ParquetFileWriter());
  result->Open(std::move(contents));
  return result;
}

const SchemaDescriptor* ParquetFileWriter::schema() const { return contents_->schema(); }

const ColumnDescriptor* ParquetFileWriter::descr(int i) const {
  return contents_->schema()->Column(i);
}

int ParquetFileWriter::num_columns() const { return contents_->num_columns(); }

int64_t ParquetFileWriter::num_rows() const { return contents_->num_rows(); }

int ParquetFileWriter::num_row_groups() const { return contents_->num_row_groups(); }

const std::shared_ptr<const KeyValueMetadata>& ParquetFileWriter::key_value_metadata()
    const {
  return contents_->key_value_metadata();
}

void ParquetFileWriter::Open(std::unique_ptr<ParquetFileWriter::Contents> contents) {
  contents_ = std::move(contents);
}

void ParquetFileWriter::Close() {
  if (contents_) {
    contents_->Close();
    contents_.reset();
  }
}

RowGroupWriter* ParquetFileWriter::AppendRowGroup() {
  return contents_->AppendRowGroup();
}

RowGroupWriter* ParquetFileWriter::AppendRowGroup(int64_t num_rows) {
  return AppendRowGroup();
}

const std::shared_ptr<WriterProperties>& ParquetFileWriter::properties() const {
  return contents_->properties();
}

}  // namespace parquet

File: rust/src/csv/writer.rs
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

use std::fs::File;
use std::io::{BufWriter, Write};

use array::*;
use datatypes::DataType;
use record_batch::RecordBatch;

pub struct Writer {
    w: BufWriter<File>,
}

macro_rules! write_primitive_array {
    ($WRITER:expr, $BATCH:expr, $ROW_INDEX:expr, $COL_INDEX:expr, $TY:ty) => {{
        let array = $BATCH
            .column($COL_INDEX)
            .as_any()
            .downcast_ref::<$TY>()
            .unwrap();
        $WRITER
            .write(format!("{}", array.value($ROW_INDEX)).as_bytes())
            .unwrap();
    }};
}

impl Writer {
    pub fn new(file: File) -> Self {
        Writer {
            w: BufWriter::new(file),
        }
    }

    pub fn write(&mut self, batch: &RecordBatch) {
        for row_index in 0..batch.num_rows() {
            for col_index in 0..batch.num_columns() {
                if col_index > 0 {
                    self.w.write(",".as_bytes()).unwrap();
                }
                match batch.schema().field(col_index).data_type() {
                    &DataType::Boolean => {
                        write_primitive_array!(self.w, batch, row_index, col_index, BooleanArray)
                    }
                    &DataType::Int8 => {
                        write_primitive_array!(self.w, batch, row_index, col_index, Int8Array)
                    }
                    &DataType::Int16 => {
                        write_primitive_array!(self.w, batch, row_index, col_index, Int16Array)
                    }
                    &DataType::Int32 => {
                        write_primitive_array!(self.w, batch, row_index, col_index, Int32Array)
                    }
                    &DataType::Int64 => {
                        write_primitive_array!(self.w, batch, row_index, col_index, Int64Array)
                    }
                    &DataType::UInt8 => {
                        write_primitive_array!(self.w, batch, row_index, col_index, UInt8Array)
                    }
                    &DataType::UInt16 => {
                        write_primitive_array!(self.w, batch, row_index, col_index, UInt16Array)
                    }
                    &DataType::UInt32 => {
                        write_primitive_array!(self.w, batch, row_index, col_index, UInt32Array)
                    }
                    &DataType::UInt64 => {
                        write_primitive_array!(self.w, batch, row_index, col_index, UInt64Array)
                    }
                    &DataType::Float32 => {
                        write_primitive_array!(self.w, batch, row_index, col_index, Float32Array)
                    }
                    &DataType::Float64 => {
                        write_primitive_array!(self.w, batch, row_index, col_index, Float64Array)
                    }
                    &DataType::Utf8 => {
                        let array = batch
                            .column(col_index)
                            .as_any()
                            .downcast_ref::<BinaryArray>()
                            .unwrap();
                        self.w.write("\"".as_bytes()).unwrap();
                        self.w.write(array.get_value(row_index)).unwrap();
                        self.w.write("\"".as_bytes()).unwrap();
                    }
                    other => panic!("unsupported type {:?}", other),
                }
            }
            self.w.write("\n".as_bytes()).unwrap();
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use datatypes::{Field, Schema};
    use std::sync::Arc;

    #[test]
    fn test_write_csv() {
        let schema = Schema::new(vec![
            Field::new("city", DataType::Utf8, false),
            Field::new("lat", DataType::Float64, false),
            Field::new("lng", DataType::Float64, false),
        ]);

        let cities = BinaryArray::from(vec![
            "Elgin, Scotland, the UK",
            "Stoke-on-Trent, Staffordshire, the UK",
            "Solihull, Birmingham, UK",
        ]);
        let lat = PrimitiveArray::from(vec![57.653484, 53.002666, 52.412811]);
        let lng = PrimitiveArray::from(vec![-3.335724, -2.179404, -1.778197]);

        let batch = RecordBatch::new(
            Arc::new(schema),
            vec![Arc::new(cities), Arc::new(lat), Arc::new(lng)],
        );

        let file = File::create("/tmp/uk_cities.csv").unwrap();

        let mut writer = Writer::new(file);
        writer.write(&batch);
    }
}
